README
# Validation

You can validate your changes to this repository by calling the validate_site.sh file with

    tests/validate_site.sh

# Dokumetation

https://gluon.readthedocs.org/en/latest/

Gluon Version auf der die Freifunk Nord Firmware basiert:

* 2017.1.1 - Gluon LEDE 2017.1.1

# Download der Firmware

* https://nord.freifunk.net/firmware.html

# Firmware selber bauen

1. Vorbereitung:
1.1 Abhängigkeiten installieren

  	sudo apt-get install git subversion build-essential gawk unzip libncurses5-dev zlib1g-dev libssl-de

1.2 Gluon repo clonen

	cd /opt/
	git clone https://github.com/freifunk-gluon/gluon


1.3 Freifunk Nord Site clonen

	cd gluon
	git clone -b nord-lede https://github.com/ffnord/nord-site.git site

2. Firmware bauen 
2.1 Build vorbereiten

	make update


2.2 Build durchführen

	make -j8 GLUON_TARGET=ar71xx-generic ##-j $ZAHL$ = Anzahl der CPU Kerne


## Mögliche Targets

ar71xx-generic
ar71xx-mikrotik
ar71xx-nand
ar71xx-tiny
mpc85xx-generic
x86-64
x86-generic
x86-geode

## Nur mit 802.11s

brcm2708-bcm2708
brcm2708-bcm2709
ipq806x
mvebu
ramips-mt7621
ramips-mt7628
ramips-rt305x
sunxi


3. Rebuild
3.1 Updaten

	cd /opt/gluon/site
	git pull
	cd ..
	git pull
	make update

3.2 Build Verzeichnis säubern

	make clean GLUON_TARGET=$TARGET


CHANGELOG.md
Änderungen mit Firmwareversion 2018.2.3 basierend auf Gluon 2018.2.3
====================================================================

Freifunk Nord spezifische Änderungen:
-------------------------------------
* In den x86-generic und x86-64 Targets sind mehr Pakete vorinstalliert: tcpdump, gre und wireguard
* Im Config Mode kann man die Koordinaten per Karte auswählen
* Wenn man den Wifi-Taster auf WiFi-Reset konfiguriert, wird jetzt auch das Gateway neu selected
* DSGVO konformer Text im Config Mode
* gluon-alt-esc Package entfernt (nicht mehr kompatibel)


Änderungen mit Firmwareversion 2017.1.8 basierend auf Gluon 2017.1.8 
======================================================================

Freifunk Nord spezifische Änderungen:
-------------------------------------
* add gluon-alt-esc Package Mit dem man direkt rausleiten kann
* verbesserter Quickfix, falls der Router abstürzt


Änderungen mit Firmwareversion 2016.2.7.2 basierend auf Gluon 2016.2.x 
======================================================================

Freifunk Nord spezifische Änderungen:
-------------------------------------
* Das Transfer-Paket und der Autoupdate-Cleaner wurden entfernt
* Es wird jetzt ein default Bandbreitenlimit gesetzt, wenn man das Feature im Config Mode aktiviert

Generelle Änderungen an Gluon 2016.2.x
--------------------------------------
* Allow upgrades to current CPE210/510 images based on upcoming Gluon master (Commit 35c9f9ae)


Änderungen mit Firmwareversion 2016.2.7.1 basierend auf Gluon 2016.2.7 
======================================================================

Freifunk Nord spezifische Änderungen:
-------------------------------------
* Das Knotenformular wurde abgeschaltet. Alle Verweise wurden aus der Firmware entfernt. Die Koordinaten und Kontaktionformationen
  werden wieder im Knoten gespeichert und können auch nur noch dort geändert werden.
* Bestehende Koordinaten und Kontaktinformationen werden mit einem Transfer-Paket vom Server geladen und im Knoten gespeichert.
  Dabei werden keine Kontaktinformationen veröffentlicht. Wer dies möchte kann seine Email Adresse, Telefonnummer
  oder Chatnamen im Config Mode seines Knoten eintragen, diese wird dann in der Knotenkarte angezeigt.
* Der Autoupdate-Cleaner bereinigt die `autoupdate.cfg` im Knoten und entfernt dabei ggf. vorhandene Altlasten.


Änderungen mit Firmwareversion 2016.2.7 basierend auf Gluon 2016.2.7 
====================================================================

Freifunk Nord spezifische Änderungen:
-------------------------------------
* gluon-quickfix überarbeitet
* Der Hardware Wifi-Taster ist beim TL-WR741 - 841 - 842N/ND und beim TL-WR1043 per default inaktiv. 
  Die Funktionalität kann im Config Mode eingestellt werden.
* Der ssid-changer checkt jetzt einmal alle 60 Minuten

Generelle Änderungen an Gluon 2016.2.7
--------------------------------------
* Verbesserter Umgang mit Fehlern beim sysupgrade (#1160)
* Backport um das kompilieren von Gluon mit Perl 5.26 oder neuer zu ermöglichen (76753ed)

Probleme mit Gluon 2016.2.7
--------------------------------------
 * Wenn Mesh on WAN deaktiviert wurde, wird bei bestimmten Modellen die MAC Adresse des WAN Ports verändert. Dies kann in Umgebungen        mit vorgeschalteten MAC Filterregeln zu Problemen führen. (#496)
 * Die TX Leistung der misten Ubiquiti Geräte ist zu hoch eingestellt. Genaue Werte sind unbekannt. Es wird empfolen die Sendeleistung 
   per Hand zu reduzieren. (#496)

Generelle Änderungen an Gluon 2016.2.6
--------------------------------------
* Sicherheitsfix CVE-2016-10229 (#1097)
* Ein Fehler beim Roamen von Clients wurde behoben (#1121)
* Ein Fehler beim kombilieren von Gluon mit OpenSSL 1.1 wurde behoben (b6a22ce)
* Ein Fehler beim kombilieren von Gluon mit langen Dateinamen wurde behoben (#1120)
* Die sysupgrade Routine wurde komplett überarbeitet (d4a69c0)

Mehr Routermodelle werden unterstützt
-------------------------------------

    ar71xx-generic
        TP-Link:    TL-WR841N/ND v12,

Probleme mit Gluon 2016.2.6
--------------------------------------
 * Wenn Mesh on WAN deaktiviert wurde, wird bei bestimmten Modellen die MAC Adresse des WAN Ports verändert. Dies kann in Umgebungen        mit vorgeschalteten MAC Filterregeln zu Problemen führen. (#496)
 * Die TX Leistung der misten Ubiquiti Geräte ist zu hoch eingestellt. Genaue Werte sind unbekannt. Es wird empfolen die Sendeleistung 
   per Hand zu reduzieren. (#496)

 Offizielle Changelogs zum nachlesen:
------------------------------------

* https://gluon.readthedocs.io/en/latest/releases/v2016.2.6.html
* https://gluon.readthedocs.io/en/latest/releases/v2016.2.7.html


Änderungen mit Firmwareversion 2016.2.5 basierend auf Gluon 2016.2.x 
================================================================

Freifunk Nord spezifische Änderungen:
-------------------------------------
* Update auf B.A.T.M.A.N. compat 15.
* Wechsel auf das MESH-Protokoll 802.11s. Neue MESH-SSID ffnord-mesh
* Nextnode Problem wurde gelößt. Neue Nextnode Adresse: 10.187.127.254.
* Neuausrichtung der netzinternen IP Vergabe. DHCP Ranges der Gateways verkleinert.
* Gateway Slots auf 20 erhöt.
* Wegfall alternativer Autoupdater Branches. Alle Nodes werden nun unter stable geführt. Das erleichtert das updaten der Knoten.
* opkg Repository liegt nun auf Gateway 0.
* Gateway VPN Keys erneuert.
* Autoupdate Server reduziert. Nun ist auch auch Public Server gelistet der von Knoten mit Uplink direkt und schneller genutzt werden   
  kann.
* rubo77 Autoupdater Signatur erneuert.
* bigfoots Autoupdater Signatur entfernt.
* Gateway Adressen geändert. Wechsel von vpnxx. auf xx.gw.
* quickfix Update auf #62e79b0 verhindert nach einem Reboot 60 Minuten einen erneuten Neustart.
* Nur Knoten die über das Knotenformular angemeldet sind können sich über VPN einwählen.
* Fehlerhaften publik IPv6 Präfix (2a03:2267:4e6f:7264) durch korekte ULA (fd42:eb49:c0b5:4242) ersetzt. 
 

tests/validate_site.sh
#!/usr/bin/env bash
set -uo pipefail

# validate_site.sh checks if the site.conf is valid json
GLUON_REPO="https://github.com/freifunk-gluon/gluon"
GLUON_BRANCH='v2019.1.x'
GLUON_PACKAGES_REPO="https://github.com/freifunk-gluon/packages"
GLUON_PACKAGES_BRANCH='v2019.1.x'

P="$(pwd)"
echo "####### check if lua5.1 is installed ..."
which lua5.1 
if [ "$?" == 1 ]; then
  echo lua5.1 not present!
  echo install with sudo apt install lua5.1
  exit 1
fi

CONFIGS="site.conf"
if [ -d "domains" ]; then
  CONFIGS="$CONFIGS "domains/*
fi

for c in $CONFIGS; do
  echo "####### validating lua $c ..."
  GLUON_SITEDIR="." GLUON_SITE_CONFIG="$c" lua5.1 tests/site_config.lua
  if [ "$?" == 1 ]; then
    exit 1
  else
    echo "OK: $c"
  fi
done
#GLUON_SITEDIR="./" GLUON_SITE_CONFIG="" lua5.1 tests/site_config.lua

echo "####### validating $P/make-release.sh ..."
bash -n "$P/make-release.sh"
if [ "$?" == 0 ]; then
  echo "OK: $P/make-release.sh"
fi

echo "####### validating $P/modules ..."
GLUON_SITE_FEEDS="none"
source "$P/modules"
testpath=/tmp/site-validate
rm -Rf "$testpath"
mkdir -p "$testpath/packages"
cd "$testpath/packages"
for feed in $GLUON_SITE_FEEDS; do
  echo "####### checking PACKAGES_${feed^^}_REPO ..."
  repo_var="PACKAGES_${feed^^}_REPO"
  commit_var="PACKAGES_${feed^^}_COMMIT"
  branch_var="PACKAGES_${feed^^}_BRANCH"
  repo="${!repo_var}"
  commit="${!commit_var}"
  branch="${!branch_var}"
  if [ "$repo" == "" ]; then
    echo "repo $repo_var missing"
    exit 1
  fi
  if [ "$commit" == "" ]; then
    echo "commit $commit_var missing"
    exit 1
  fi
	if [ "$branch" == "" ]; then
    echo "branch $branch_var missing"
    exit 1
  fi
  git clone -b "$branch" --depth 1000 --single-branch "$repo" "$feed"
  if [ "$?" != "0" ]; then exit 1; fi
  cd "$feed"
  echo "git checkout $commit"
  git checkout "$commit"
  if [ "$?" != "0" ]; then exit 1; fi
  cd -
done

echo "####### Lua linter check for all package feeds ..."
~/.luarocks/bin/luacheck --config "$P/tests/.luacheckrc" "$testpath/packages"

echo "####### downloading $GLUON_PACKAGES_REPO ..."
git clone -b "$GLUON_PACKAGES_BRANCH" --depth 1  --single-branch "$GLUON_PACKAGES_REPO"

echo "####### downloading gluon ..."
cd "$testpath"
git init gluon
cd gluon
git remote add origin "$GLUON_REPO"
git config core.sparsecheckout true
echo "package/*" >> .git/info/sparse-checkout
git pull --depth 1 origin "$GLUON_BRANCH"
cp -a package/ "$testpath/packages"
cd "$testpath/packages/package"

echo "####### validating GLUON_SITE_PACKAGES from $P/site.mk ..."
# ignore non-gluon packages and standard gluon features
sed '0,/^GLUON_LANGS/d' "$P/site.mk" | sed '/GLUON_TARGET/,$d' | egrep -v '(#|G|iwinfo|iptables|haveged|vim|socat|tar|mesh-batman-adv-1[45]|web-advanced|web-wizard)'> "$testpath"/site.mk.sh
sed -i 's/\s\\$//g;/^$/d' "$testpath"/site.mk.sh
sed -i 's/gluon-mesh-batman-adv-1[45]/gluon-mesh-batman-adv/g' "$testpath"/site.mk.sh
sed -i 's/gluon-config-mode-geo-location-with-map/gluon-config-mode-geo-location/g' $testpath/site.mk.sh
cat "$testpath"/site.mk.sh |
while read packet; do
  if [ "$packet" != "" ]; then
    echo -n "# $packet"
    FOUND="$(find "$testpath/packages/" -type d -name "$packet")"
    if [ "$FOUND" '!=' '' ]; then
      echo " found in $(echo "$FOUND"|sed 's|'"$testpath/packages"'||g')"
    else
      # check again with prefix gluon-
      FOUND="$(find "$testpath/packages/" -type d -name "gluon-$packet")"
      if [ "$FOUND" '!=' '' ]; then
        echo " found as FEATURE in $(echo "$FOUND"|sed 's|'"$testpath/packages"'||g')"
      else
        echo
        echo "ERROR: $packet missing"
        exit 1
      fi
    fi
  fi
done


.gitlab-ci/build.sh
#!/bin/bash -e
# =====================================================================
# Build script for Freifunk Nord firmware
#
# Source: 
# Contact: 
# Web: 
#
# Credits:
#   - Freifunk Darmstadt for your great support
#   - Freifunk Fulda for the base of the gitlab-ci support
# =====================================================================

# Default make options (override with -m)
MAKEOPTS="-j 4 GLUON_DEPRECATED=upgrade"

# Default is set to use current work directory
SITEDIR="$(pwd)"

# Default build identifier set to snapshot
BUILD="snapshot"

# Specify deployment server and user
DEPLOYMENT_SERVER="10.116.250.1"
DEPLOYMENT_USER="rsync"
DEPLOYMENT_PATH="/opt/firmware/nord"

# Path to signing key
SIGNKEY=""

# Error codes
E_ILLEGAL_ARGS=126

# Help function used in error messages and -h option

usage() {
  echo ""
  echo "Build script for Freifunk-Fulda gluon firmware."
  echo ""
  echo "-a: Autoupdater branch name (e.g. development)"
  echo "    Default: branch (see -b)"
  echo "-b: Firmware branch name (e.g. development)"
  echo "    Default: current git branch"
  echo "-c: Build command: update | clean | download | build | sign | upload | prepare"
  echo "-d: Enable bash debug output"
  echo "-h: Show this help"
  echo "-m: Setting for make options (optional)"
  echo "    Default: \"${MAKEOPTS}\""
  echo "-n: Build identifier (optional)"
  echo "    Default: \"${BUILD}\""
  echo "-t: Gluon targets architectures to build"
  echo "    Default: \"${TARGETS}\""
  echo "-u: Upload target"
  echo "    Default: branch (see -b)"
  echo "-r: Release number (optional)"
  echo "    Default: fetched from release file"
  echo "-w: Path to site directory"
  echo "    Default: current working directory"
  echo "-s: Path to signing key"
  echo "    Default: empty"
}

# Evaluate arguments for build script.
if [[ "${#}" == 0 ]]; then
  usage
  exit ${E_ILLEGAL_ARGS}
fi

# Evaluate arguments for build script.
while getopts a:b:c:dhm:n:t:u:w:s: flag; do
  case ${flag} in
    a)
        AU_BRANCH="${OPTARG}"
        ;;
    b)
        BRANCH="${OPTARG}"
	if [ "${BRANCH}" == "release-candidate" ] ; then
          BRANCH="rc"
	fi
        ;;
    c)
      case "${OPTARG}" in
        update)
          COMMAND="${OPTARG}"
          ;;
        clean)
          COMMAND="${OPTARG}"
          ;;
        download)
          COMMAND="${OPTARG}"
          ;;
        build)
          COMMAND="${OPTARG}"
          ;;
        sign)
          COMMAND="${OPTARG}"
          ;;
        upload)
          COMMAND="${OPTARG}"
          ;;
        prepare)
          COMMAND="${OPTARG}"
          ;;
        *)
          echo "Error: Invalid build command set."
          usage
          exit ${E_ILLEGAL_ARGS}
          ;;
      esac
      ;;
    d)
      set -x
      ;;
    h)
      usage
      exit
      ;;
    m)
      MAKEOPTS="${OPTARG}"
      ;;
    n)
      BUILD="${OPTARG}"
      ;;
    t)
      TARGETS="${OPTARG}"
      ;;
    u)
      UPLOAD_TARGET="${OPTARG}"
      ;;
    r)
      RELEASE="${OPTARG}"
      ;;
    w)
      # Use the root project as site-config for make commands below
      SITEDIR="${OPTARG}"
      ;;
    s)
      SIGNKEY="${OPTARG}"
      ;;
    *)
      usage
      exit ${E_ILLEGAL_ARGS}
      ;;
  esac
done

# Strip of all remaining arguments
shift $((OPTIND - 1));

# Check if there are remaining arguments
if [[ "${#}" > 0 ]]; then
  echo "Error: Too many arguments: ${*}"
  usage
  exit ${E_ILLEGAL_ARGS}
fi

# Set branch name
if [[ -z "${BRANCH}" ]]; then
  BRANCH=$(git symbolic-ref -q HEAD)
  BRANCH=${BRANCH##refs/heads/}
  BRANCH=${BRANCH:-HEAD}
fi

# Default to build branch specific targets if parameter -t is not set 
if [[ -z ${TARGETS+x} ]] ; then
  case "${BRANCH}" in
    stable)
      TARGETS="ar71xx-tiny ar71xx-generic"
      TARGETS+=" x86-64 x86-generic" # (VMs)
      TARGETS+=" ar71xx-nand" # (Netgear WNDR3700, WNDR4300, ZyXEL NBG6716)
      #TARGETS+=" mpc85xx-generic" # (tp-link-tl-wdr4900-v1)
      TARGETS+=" ramips-mt7620" # (gl-inet mt300 und mt750)
      #TARGETS+=" sunxi-cortexa7" # (Banana Pi M1)

      # BROKEN:
      #TARGETS+=" brcm2708-bcm2708 brcm2708-bcm2709 brcm2708-bcm2710" # (raspberry Pi 1, 2 und 3)
      TARGETS+=" ipq40xx" # (FitzBox 4040)
      #TARGETS+=" ramips-mt7621" # (D-Link DIR-860L (B1) Ubiquiti EdgeRouter X, ZBT WG3526)
      TARGETS+=" x86-geode"
      TARGETS+=" ramips-rt305x" # BROKEN: (fonera, vocore a5)
      TARGETS+=" ramips-mt76x8" # BROKEN: unstable WiFi (tp-link 841 v13, Netgear R6120 und archer c50)
      #TARGETS+=" ar71xx-mikrotik" # BROKEN: no sysupgrade support (mikrotik-nand)
      #TARGETS+=" brcm2708-bcm2710" # BROKEN: Untested (raspberry-pi-3)
      #TARGETS+=" ipq806x" # BROKEN: unstable wifi drivers (tp-link-archer-c2600)
      #TARGETS+=" mvebu-cortexa9" # BROKEN: No AP+IBSS or 11s support (linksys-wrt1200ac)
      ;;
    *)
      # Default to all targets
      TARGETS="ar71xx-generic ar71xx-tiny ar71xx-nand brcm2708-bcm2708 brcm2708-bcm2709 ramips-mt7621 x86-generic x86-geode x86-64 ramips-mt7620 ramips-mt76x8 ramips-rt305x sunxi-cortexa7 ipq40xx"
      TARGETS+=" mpc85xx-generic" # (tp-link-tl-wdr4900-v1)-
    ;;
  esac
fi

if [[ -z "$AU_BRANCH" ]]; then
  AU_BRANCH="$BRANCH"
fi

if [[ -z "$UPLOAD_TARGET" ]]; then
  UPLOAD_TARGET="$BRANCH"
fi

# Set command
if [[ -z "${COMMAND}" ]]; then
  echo "Error: Build command missing."
  usage
  exit ${E_ILLEGAL_ARGS}
fi

# Normalize the branch name
BRANCH="${BRANCH#origin/}" # Use the current git branch as autoupdate branch
BRANCH="${BRANCH//\//-}"   # Replace all slashes with dashes

# Set release number
if [[ -z "${RELEASE}" ]]; then
  RELEASE=$(sed -e "s/BUILD/$BUILD/" "${SITEDIR}/release")
  case "${BRANCH}" in
    nightly)
      RELEASE="${RELEASE}~ngly"
      ;;
    next)
      RELEASE="${RELEASE}~next"
      ;;
    *)
      # Do nothing
      ;;
  esac
fi

# Get the GIT commit description
COMMIT="$(git describe --always --dirty)"

# Number of days that may pass between releasing an updating
if [[ -z ${PRIORITY+x} ]] ; then
  case "${BRANCH}" in
    nightly)
      PRIORITY=0
      ;;
    *)
      PRIORITY=1
      ;;
  esac
fi

update() {
  echo "--- Update Gluon Dependencies"
  make ${MAKEOPTS} \
       GLUON_SITEDIR="${SITEDIR}" \
       GLUON_OUTPUTDIR="${SITEDIR}/output" \
       GLUON_RELEASE="${RELEASE}" \
       GLUON_BRANCH="${AU_BRANCH}" \
       GLUON_PRIORITY="${PRIORITY}" \
       update
}

clean() {
  for TARGET in ${TARGETS}; do
    echo "--- Clean Gluon Build Artifacts for target: ${TARGET}"
    make ${MAKEOPTS} \
         GLUON_SITEDIR="${SITEDIR}" \
         GLUON_OUTPUTDIR="${SITEDIR}/output" \
         GLUON_RELEASE="${RELEASE}" \
         GLUON_BRANCH="${AU_BRANCH}" \
         GLUON_PRIORITY="${PRIORITY}" \
         GLUON_TARGET="${TARGET}" \
         clean
  done
}

download() {
  for TARGET in ${TARGETS}; do
    echo "--- Download Gluon Dependencies for target: ${TARGET}"
    make ${MAKEOPTS} \
         GLUON_SITEDIR="${SITEDIR}" \
         GLUON_OUTPUTDIR="${SITEDIR}/output" \
         GLUON_RELEASE="${RELEASE}" \
         GLUON_BRANCH="${AU_BRANCH}" \
         GLUON_PRIORITY="${PRIORITY}" \
         GLUON_TARGET="${TARGET}" \
         download
  done
}

build() {
  for TARGET in ${TARGETS}; do
    echo "--- Build Gluon Images for target: ${TARGET}"
    case "${AU_BRANCH}" in
      stable)
        make ${MAKEOPTS} \
             GLUON_SITEDIR="${SITEDIR}" \
             GLUON_OUTPUTDIR="${SITEDIR}/output" \
             GLUON_RELEASE="${RELEASE}" \
             GLUON_BRANCH="${AU_BRANCH}" \
             GLUON_PRIORITY="${PRIORITY}" \
             GLUON_TARGET="${TARGET}"
        ;;

      *)
        make ${MAKEOPTS} \
             GLUON_SITEDIR="${SITEDIR}" \
             GLUON_OUTPUTDIR="${SITEDIR}/output" \
             GLUON_RELEASE="${RELEASE}" \
             GLUON_BRANCH="${AU_BRANCH}" \
             GLUON_TARGET="${TARGET}"
      ;;
    esac
  done

  echo "--- Build Gluon Manifest"
  make ${MAKEOPTS} \
       GLUON_SITEDIR="${SITEDIR}" \
       GLUON_OUTPUTDIR="${SITEDIR}/output" \
       GLUON_RELEASE="${RELEASE}" \
       GLUON_BRANCH="${BRANCH}" \
       GLUON_PRIORITY="${PRIORITY}" \
       manifest

  cp "${SITEDIR}/output/images/sysupgrade/${BRANCH}.manifest" \
     "${SITEDIR}/output/images/sysupgrade/${BRANCH}.manifest.clean"

  echo "--- Write Build file"
  cat > "${SITEDIR}/output/images/build" <<EOF
DATE=$(date '+%Y-%m-%d %H:%M:%S')
VERSION=$(cat "${SITEDIR}/release")
RELEASE=${RELEASE}
BUILD=${BUILD}
BRANCH=${BRANCH}
COMMIT=${COMMIT}
HOST=$(uname -n)
EOF
}

sign() {
  echo "--- Sign Gluon Firmware Build"

  # Add the signature to the local manifest
  contrib/sign.sh \
      "${SIGNKEY}" \
      "${SITEDIR}/output/images/sysupgrade/${BRANCH}.manifest"
}

upload() {
  set -x
  echo "--- Upload Gluon Firmware Images and Manifest"

  # Build the ssh command to use
  SSH="ssh"
  SSH="${SSH} -o stricthostkeychecking=no -v"

  # Determine upload target prefix
  TARGET="${UPLOAD_TARGET}"

  # Create the target directory on server
  ${SSH} \
      ${DEPLOYMENT_USER}@${DEPLOYMENT_SERVER} \
      -- \
      mkdir \
          --parents \
          --verbose \
          "${DEPLOYMENT_PATH}/${TARGET}/${RELEASE}"

  # Add site metadata
  tar -czf "${SITEDIR}/output/images/site.tgz" --exclude='gluon' --exclude='output' "${SITEDIR}"

  # Compress images (Saves around 40% space, relevant because of shitty VDSL 50 upload speeds)
  echo "Compressing images..."
  tar -cJf "${SITEDIR}/output/images.txz" -C "${SITEDIR}/output/images/" factory sysupgrade other

  # Copy images to server
  echo "Uploading images..."
  rsync \
      --verbose \
      --progress \
      --chmod=ugo=rwX \
      --rsh="${SSH}" \
      "${SITEDIR}/output/images.txz" \
      "${DEPLOYMENT_USER}@${DEPLOYMENT_SERVER}:${DEPLOYMENT_PATH}/${TARGET}/${RELEASE}"

  echo "Uncompressing images..."
  ${SSH} \
      ${DEPLOYMENT_USER}@${DEPLOYMENT_SERVER} \
      -- \
     tar -xJf "${DEPLOYMENT_PATH}/${TARGET}/${RELEASE}/images.txz" -C "${DEPLOYMENT_PATH}/${TARGET}/${RELEASE}/"

  ${SSH} \
      ${DEPLOYMENT_USER}@${DEPLOYMENT_SERVER} \
      -- \
      ln -sf \
          "${DEPLOYMENT_PATH}/${TARGET}/${RELEASE}/sysupgrade" \
          "${DEPLOYMENT_PATH}/${TARGET}/"
  ${SSH} \
      ${DEPLOYMENT_USER}@${DEPLOYMENT_SERVER} \
      -- \
      ln -sf \
          "${DEPLOYMENT_PATH}/${TARGET}/${RELEASE}/factory" \
          "${DEPLOYMENT_PATH}/${TARGET}/"
  ${SSH} \
      ${DEPLOYMENT_USER}@${DEPLOYMENT_SERVER} \
      -- \
      ln -sf \
          "${DEPLOYMENT_PATH}/${TARGET}/${RELEASE}/other" \
          "${DEPLOYMENT_PATH}/${TARGET}/"
}

prepare() {
  echo "--- Prepare directory for upload"

  # Determine upload target prefix
  TARGET="${UPLOAD_TARGET}"

  # Create the target directory on server
  mkdir \
    --parents \
    --verbose \
    "${SITEDIR}/output/firmware/${TARGET}"

  # Copy images to directory
  mv \
    --verbose \
    "${SITEDIR}/output/images" \
    "${SITEDIR}/output/firmware/${TARGET}/${RELEASE}"

  # Link latest upload in target to 'current'
  cd "${SITEDIR}/output"
  ln \
      --symbolic \
      --force \
      --no-target-directory \
      "${RELEASE}" \
      "firmware/${TARGET}/current"
}

(
  # Change working directory to gluon tree
  cd "${SITEDIR}/gluon"

  # Execute the selected command
  ${COMMAND}
)


make-release.sh
#!/bin/bash

## This script will compile Gluon for all architectures, create the
## manifest and sign it. For that, you must have cloned gluon and have a
## valid site config. Additionally, the signing key must be present in
## ../../ecdsa-key-secret or defined as first argument.
## The second argument defines the branch (stable, rc, nightly).
## The third argument defines the version.
## Call from site directory with the version and branch variables
## properly configured in this script.

# report the usage of uninitialized variables, etc
set -u
# stop on first error (not working in if clauses)
set -e

# if version is unset, will use the default version from site.mk
#VERSION=${3:-"2018.2.1~exp$(date '+%y%m%d%H%M')"}
VERSION=${3:-"2018.2.1"}
# branch must be set to either rc, nightly or stable
BRANCH=${2:-"stable"}
# must point to valid ecdsa signing key created by ecdsakeygen, relative to Gluon base directory
SIGNING_KEY=${1:-"../ecdsa-key-secret"}

# BROKEN must be set to "" or "BROKEN=1"
BROKEN="BROKEN=1"

# set num cores +1
CORES=$(($(lscpu|grep -e '^CPU(s):'|xargs|cut -d" " -f2)+1))
CORES="-j$CORES"

# set this to "0" if you don't want to use make clean before make
MAKE_CLEAN="0"

# set this to "" to get less more output
VERBOSE="V=s"

#ONLY_TARGET must be set to "" or i.e. "ar71xx-tiny"
#ONLY_TARGET=""
ONLY_TARGET="ar71xx-generic"
#to build only one device set DEVICES list (only if $ONLY_TARGET!="")
#DEVICES=''
DEVICES='DEVICES=tp-link-tl-wr842n-nd-v3'

cd ../
if [ ! -d "site" ]; then
  echo "This script must be called from within the site directory"
  exit
fi

if [ "$(whoami)" == "root" ]; then
  echo "Make may not be run as root"
fi

echo "############## starting build process #################" >> build.log
date >> build.log
echo "if you want to start over empty the folder ../output/"
echo "see debug output with"
echo "tail -F ../build.log|grep -i error|grep -v CFLAGS|egrep -v '(checking|CC|LD|gcc|Entering|leaving|v -f|rm -rf|cp -f|rm -f|Applying|patching|Installing)' &"
sleep 3

#rm -r output

#  ramips-mt7621:  BROKEN: No AP+IBSS support, 11s has high packet loss
#  ramips-rt305x:  BROKEN: No AP+IBSS support

WRT1200AC="mvebu" # Linksys WRT1200AC BROKEN: No AP+IBSS+mesh support

ONLY_11S="ramips-rt305x ramips-mt7621"    # BROKEN only

ONLY_LEDE="ar71xx-tiny" # Support for for 841 on lede, needs less packages, so the 4MB will suffice!
ONLY_LEDE+=" x86-geode ipq806x ramips-mt76x8"
NOT_LEDE="x86-kvm_guest" # The x86-kvm_guest target has been dropped from LEDE; x86-64 should be used

BANANAPI="sunxi-cortexa7"                          # BROKEN: Untested, no sysupgrade support
MICROTIK="ar71xx-mikrotik"                # BROKEN: no sysupgrade support

RASPBPI="brcm2708-bcm2708 brcm2708-bcm2709"
X86="x86-64 x86-generic x86-xen_domu"
WDR4900="mpc85xx-generic"

TARGETS="ar71xx-generic $ONLY_LEDE ar71xx-nand $WDR4900 $RASPBPI $X86"
if [ "$BROKEN" != "" ]; then
  TARGETS+=" $BANANAPI $MICROTIK $WRT1200AC"
fi

if [ "$ONLY_TARGET" != "" ]; then
  TARGETS="$ONLY_TARGET"
fi

for TARGET in $TARGETS; do
  date >> build.log
  echo "Starting work on target $TARGET $DEVICES" | tee -a build.log
  OPTIONS="GLUON_TARGET=$TARGET $BROKEN $CORES GLUON_BRANCH=$BRANCH GLUON_RELEASE=$VERSION"
  echo -e "\n===========\n\n\n\n\nmake $OPTIONS update" >> build.log
  time make $OPTIONS update >> build.log 2>&1
  if [ $MAKE_CLEAN = 1 ]; then
    echo -e "\n===========\n\n\n\n\nmake $OPTIONS clean" >> build.log
    time make $OPTIONS clean >> build.log 2>&1
  fi
  echo -e "\n===========\n\n\n\n\nmake $OPTIONS $DEVICES $VERBOSE" >> build.log
  time make $OPTIONS $DEVICES $VERBOSE >> build.log 2>&1
  echo -e "\n\n\n============================================================\n\n" >> build.log
done
date >> build.log

echo "Compilation complete, creating manifest(s)" | tee -a build.log
set +e
MANIFEST_OPTINS="GLUON_RELEASE=$VERSION $BROKEN $CORES"
if [[ true ]]; then
  B="nightly"
  echo -e "make $MANIFEST_OPTINS GLUON_BRANCH=$B manifest" >> build.log
  make $MANIFEST_OPTINS GLUON_BRANCH=$B manifest >> build.log 2>&1
  echo -e "\n\n\n============================================================\n\n" >> build.log
fi

if [[ "$BRANCH" == "stable" ]]; then
  B="stable"
  echo -e "make $MANIFEST_OPTINS GLUON_BRANCH=$B manifest" >> build.log
  make $MANIFEST_OPTINS GLUON_BRANCH=$B manifest >> build.log 2>&1
  echo -e "\n\n\n============================================================\n\n" >> build.log
fi

echo "Manifest creation complete, signing manifest"

echo -e "contrib/sign.sh $SIGNING_KEY output/images/sysupgrade/nightly.manifest" >> build.log
contrib/sign.sh $SIGNING_KEY output/images/sysupgrade/nightly.manifest >> build.log 2>&1

if [[ "$BRANCH" == "nightly" ]] || [[ "$BRANCH" == "stable" ]]; then
  echo -e "contrib/sign.sh $SIGNING_KEY output/images/sysupgrade/nightly.manifest" >> build.log
  contrib/sign.sh $SIGNING_KEY output/images/sysupgrade/nightly.manifest >> build.log 2>&1
  # set date to before 04:00
  sed -e 's/DATE=.*/DATE='$(date '+%y-%m-%d')' 00:00:00+02:00/g' output/images/sysupgrade/nightly.manifest
fi

if [[ "$BRANCH" == "stable" ]]; then
  echo -e "contrib/sign.sh $SIGNING_KEY output/images/sysupgrade/stable.manifest" >> build.log
  contrib/sign.sh $SIGNING_KEY output/images/sysupgrade/stable.manifest >> build.log 2>&1
fi
cd site
date >> ../build.log
mv -v ../output/images "../output/$VERSION"
echo "Done :)"
